### ISSUE TITLE
Problem with secondary types

### ISSUE THREAD

**Issue by bjitivo** [back2dos/single_impl#1]
Thank you for this macro.  I think it's almost working, but the first test case I tried revealed a problem:

Secondary types within a .hx file appear to have an "_" prepended to them, at least for the hxcpp target, and this underscore ends up in the name of the interface to apply the Skip macro to, which then does not match the actual interface name when the compiler is run.

**Comment by bjitivo** [comment]
Also, things don't seem to work quite right.  It would appear that uses of the removed interface type in Std.is() and in function arguments ends up getting the "Nada" class, which is wrong.  They should get the single-implementor class.

**Comment by bjitivo** [comment]
It appears that this latter problem is because of the "crude" detection logic in Skip.hx; turns out that this is _also_ a multi-line "call site":

private function foo(i : SinglyImplementedInterface)
{
}

**Comment by back2dos** [comment]
Yeah, there is a better macro based solution for all of this, that should
do it in a single pass. I'll see if I can find the time to try it this
weekend.

**Comment by bjitivo** [comment]
Thank you ... I believe I have a version that is basically working for us, there is one remaining issue though, and that is that the haxe compiler doesn't call back generic macros for Std.is() and similar constructs.  So this whole approach can't even work properly until the compiler authors fix that.

Thank you very much for leading the way, in any case.

**Comment by back2dos** [comment]
Module subtypes now work for me, although I cannot confirm the "_" surfacing anywhere. In the sample code, [`example.A.B`](https://github.com/back2dos/single_impl/blob/master/src/example/A.hx#L11) is simply generated as `bin/example/B.cpp`. Do you have a minimal example illustrating the problem?

**Comment by bjitivo** [comment]
OK here's a test that fails:

<pre>
class Lonely
{
    public static function main()
    {
        var i : ILonelyInterface = new LonelyImplementation();
        i.doSomething();
    }
    static function doAThing(i : ILonelyInterface) : Int
    {
        i.doSomething();
        return 100;
    }
}
private interface ILonelyInterface
{
    function doSomething() : Void;
}
private class LonelyImplementation implements ILonelyInterface
{
    public function doSomething() : Void
    {
        trace("Hello, world!");
    }
    public function new()
    {
    }
}
</pre>

Output:

<pre>
Error: In file included from ./src/Lonely.cpp:4:
include/Lonely.h:35: error: '_Lonely' has not been declared
include/Lonely.h:35: error: expected ',' or '...' before 'i'
./src/Lonely.cpp:45: error: prototype for 'int Lonely_obj::doAThing(_Lonely::LonelyImplementation)' does not match any in class 'Lonely_obj'
include/Lonely.h:35: error: candidate is: static int Lonely_obj::doAThing(int)
Error: Build failed
</pre>

**Comment by bjitivo** [comment]
Seems to be something with the hxcpp backend not generating an include for "_Loney/LonelyImpementation.h" for some reason ...

**Comment by bjitivo** [comment]
OK, digging further, it would appear that the hxcpp backend knows to emit these into the header file for any type referenced as a function argument:

HX_DECLARE_CLASS1(_Lonely,JustSomeClass)

This is in lieu of actually #including the header file that defines that class.

However, this HX_DECLARE_CLASS line is _not_ emitted for classes that have replaced interfaces that were removed, when those removed interfaces are used as function argument types.

**Comment by bjitivo** [comment]
If I remove the "x.orphan.exclude()" line from the Sweep.use() function, then the proper HX_DECLARE_CLASS invocation ends up in the Lonely.h header file.  But then LonelyImplementation.o ends up being included twice on the all_objs file, which leads to a double-inclusion when linking, which causes duplicate symbol errors ...

**Comment by bjitivo** [comment]
Is this a bug with how @:native is treated by the hxcpp back end?  It doesn't include the rewritten type as one to forward-declare, not sure why not ...

**Comment by back2dos** [comment]
Yeah, this is weird. It only seems to cause issues for method arguments,
but it works for variable types and even for local functions. Go figure ...
I skipped module subtypes for now. My non-existent C++ skills don't exactly
help here :D​

**Comment by bjitivo** [comment]
Not an ocaml expert, but if I change the arguments to find_referenced_types in gencpp.ml from:

  (\* Only need to foreward-declare classes that are mentioned in the header file
      (ie, not the implementation)  *)
   let referenced = find_referenced_types ctx.ctx_common (TClassDecl class_def) super_deps (Hashtbl.create 0) true false scriptable in
   List.iter ( gen_forward_decl h_file ) referenced;

to 

  (\* Only need to foreward-declare classes that are mentioned in the header file
      (ie, not the implementation)  *)
   let referenced = find_referenced_types ctx.ctx_common (TClassDecl class_def) super_deps (Hashtbl.create 0) false false scriptable in
   List.iter ( gen_forward_decl h_file ) referenced;

Then things work.  I don't fully understand why though ...

**Comment by back2dos** [comment]
Wow. This hurts my brain. But if it works, it works. Have you tried your
magic fix against the actual code base?

**Comment by bjitivo** [comment]
I'm trying it now.  If it works, I will submit an issue for hxcpp asking why the find_referenced_types function thinks it can ignore fields and such when looking for types to forward declare in header files, that seems wrong to me, and my hack prevents it, which seems to fix the problem.

**Comment by bjitivo** [comment]
Argh.  This fails too, even with my "fix":

<pre>
class Lonely
{
    public static function main()
    {
    }
    static var mI : ILonelyInterface = null;
}
private interface ILonelyInterface
{
    function doSomething() : Void;
}
private class LonelyImplementation implements ILonelyInterface
{
    public function doSomething() : Void
    {
        trace("Hello, world!");
    }
    public function new()
    {
    }
}
</pre>

Result:

<pre>
Error: In file included from ./src/Lonely.cpp:4:
include/Lonely.h:37: error: '_Lonely' has not been declared
include/Lonely.h:37: error: ISO C++ forbids declaration of 'LonelyImplementation' with no type
include/Lonely.h:37: error: expected ';' before 'mI'
./src/Lonely.cpp:38: error: '_Lonely' has not been declared
./src/Lonely.cpp:38: error: expected constructor, destructor, or type conversion before 'Lonely_obj'
./src/Lonely.cpp: In static member function 'static bool Lonely_obj::__GetStatic(const String&, Dynamic&, hx::PropertyAccess)':
./src/Lonely.cpp:49: error: 'mI' was not declared in this scope
./src/Lonely.cpp: In static member function 'static bool Lonely_obj::__SetStatic(const String&, Dynamic&, hx::PropertyAccess)':
./src/Lonely.cpp:61: error: 'mI' was not declared in this scope
./src/Lonely.cpp:61: error: '_Lonely' was not declared in this scope
./src/Lonely.cpp:61: error: parse error in template argument list
./src/Lonely.cpp:61: error: no matching function for call to 'Dynamic::Cast()'
./src/Lonely.cpp: In function 'void sMarkStatics(hx::MarkContext*)':
./src/Lonely.cpp:76: error: 'mI' is not a member of 'Lonely_obj'
./src/Lonely.cpp: In static member function 'static void Lonely_obj::__boot()':
./src/Lonely.cpp:122: error: 'mI' was not declared in this scope
</pre>

**Comment by back2dos** [comment]
Ok, I see. Well, if you can do without elimination of module subtypes, then
my last changes should do the trick. But it seems like Simn is converging
towards a solution anyway.

**Comment by bjitivo** [comment]
Well, I like your macro better ... it doesn't require weird compiler changes.

**Comment by back2dos** [comment]
Hehe, ok. Well, how much of an issue is it to leave module subtypes
untouched?

It this is important, then I could use some help with the C++ part. Am I
correct in assuming that the issue is that some include statement is not
generated? If so, maybe a `@:buildXml` or `@:include` in the right place
could do the trick.

**Comment by bjitivo** [comment]
OK, so I figured out the problem with hxcpp.  The problem is that the call to x.orphan.exclude() is a call whose implementation just sets the "cl_extern" field of the class in the compiler.  However, the hxcpp backend doesn't ever generate the forward references for classes with "cl_extern" set.  I don't know why.

So I changed the compiler to also keep a flag "cl_excluded" which I use to detect this situation and overcome it.  It's a hack and I am sure there is a correct solution, but I will have to confer with Hugh of hxcpp on that.

**Comment by bjitivo** [comment]
Here's an example that doesn't work with the single_impl macro:

<pre>
class Lonely
{
    public static function main()
    {
        var getter : IGetter = new Getter();
        var bad : ILonely = getter.get();
    }
}
interface ILonely
{
}
interface ILonelySub extends ILonely
{
}
interface IGetter
{
    function get() : ILonely;
}
class Getter implements IGetter
{
    public function get() : ILonelySub
    {
        return null;
    }
    public function new()
    {
    }
}
class LonelyImplementor implements ILonely
{
}
</pre>

The problem here is that:

ILonely is replaced by LonelyImplementor since it's the only implementor of that interface.  IGetter is replaced by Getter since it's the only implementation of that interface.

So the resulting code is:

<pre>
class Lonely
{
    public static function main()
    {
        var getter : Getter = new Getter();
        var bad : LonelyImplementor = getter.get();
    }
}
interface ILonelySub
{
}
class Getter
{
    public function get() : ILonelySub
    {
        return null;
    }
    public function new()
    {
    }
}
class LonelyImplementor
{
}
</pre>

The code that is generated thus tries to assign an instance of ILonelySub returned by Getter.get() to a variable of type LonelyImplementor; there is no longer any relationship between these two types.  The C++ code that is generated refuses to compile ...

**Comment by back2dos** [comment]
Ok, that should be covered now as well. Such completely lonely interfaces just get replaced by Dynamic now. I hope that's good enough given that `null` is the only valid value for them.

**Comment by bjitivo** [comment]
Nah, still doesn't work for the cpp target.  neko is fine though.

You can test this by moving all of the classes and interfaces in my most recent example into their own files so they are not "subtypes", then compiling with: haxe -cp single_impl/src --macro "single_impl.Sweep.use()" -main Lonely -cpp out.
